File: src/purra/agent_tree_tool.py
"""Independent model tools over Agent commands, queries and result reception."""
from __future__ import annotations

import json
from dataclasses import dataclass

from purra.agent_tree import ChildAgentSpec, ContinueAgentCommand, SpawnAgentsCommand
from purra.agent_tree_execution import RunCommandService
from purra.agent_tree_policy import AgentTreePolicy
from purra.contracts import ToolExecutionMode, ToolHandlerResult, ToolPlanningDisposition, ToolPolicy, ToolRiskLevel, ToolSchema
from purra.errors import ContractViolationError
from purra.json_values import thaw_json_mapping
from purra.ports import ToolRegistration


@dataclass(frozen=True)
class AgentToolContext:
    commands: RunCommandService
    policy: AgentTreePolicy
    child_allowed_tools: tuple[str, ...] = ()
    lease_owner_id: str | None = None
    lease_epoch: int | None = None

    @property
    def claim(self):
        return {"lease_owner_id": self.lease_owner_id, "lease_epoch": self.lease_epoch}

    async def receive(self, run_id, ids, signal, *, after=()):
        aggregate = await self.commands.receive_runs(run_id, tuple(ids), signal, after_run_ids=tuple(after), **self.claim)
        # Result reception describes only the Agents participating in this call.
        agents = []
        seen = set()
        for identity in ids:
            agent_id = await self.commands.agents.agent_id_for_run(run_id, identity)
            if agent_id not in seen:
                agents.append(await self.commands.agents.describe(run_id, agent_id))
                seen.add(agent_id)
        return ToolHandlerResult(
            json.dumps({"state": aggregate.state, "agents": agents,
                "runIds": ids, "pendingRunIds": list(aggregate.pending_run_ids),
                "requiredFailures": list(aggregate.required_failures),
                "results": [thaw_json_mapping(item) for item in aggregate.results]}, ensure_ascii=False, separators=(",", ":")),
            error_code="required_child_run_failed" if aggregate.state == "blocked" else None,
            planning_disposition=ToolPlanningDisposition.REPLAN,
        )


def _schema(properties, required=()):
    return {"type": "object", "properties": properties, "required": list(required), "additionalProperties": False}


def _registration(name, title, description, parameters, handler, *, read=False):
    async def unavailable(state, arguments, signal=None):
        raise ContractViolationError("Agent tool requires ToolCall identity")

    async def handle(state, arguments, tool_call, signal=None):
        run_id = str(state.run_id or "").strip()
        if not run_id:
            raise ContractViolationError("Agent tool requires a bound Run")
        return await handler(run_id, arguments, tool_call, signal)

    return ToolRegistration(
        schema=ToolSchema(name=name, display_names={"zh-CN": title, "en-US": name}, description=description, parameters=parameters),
        handler=unavailable, call_handler=handle,
        policy=ToolPolicy(mode=ToolExecutionMode.READ if read else ToolExecutionMode.PROPOSE,
                          title=title, risk_level=ToolRiskLevel.READ if read else ToolRiskLevel.WRITE),
        host_managed_durability=True,
    )


def _delegate(context):
    policy = context.policy
    async def handle(run_id, arguments, call, signal):
        items = policy.validate_children(arguments.get("children"))
        grant = await context.commands.compile_child_grant(run_id, can_spawn_agents=policy.allow_recursive_agents, allowed_tools=context.child_allowed_tools)
        receipt = await context.commands.spawn_agents(SpawnAgentsCommand(
            parent_run_id=run_id, idempotency_key=call.id, **context.claim,
            children=tuple(ChildAgentSpec(name=item["name"], title=item["title"], instruction=item["instruction"],
                objective=item["objective"], input_payload=item["input"], required=item["required"],
                priority=item["priority"], capability_grant=grant) for item in items)))
        return await context.receive(run_id, [item.run.run_id for item in receipt.items], signal)

    item = _schema({
        key: {"type": "string", "minLength": 1, "maxLength": maximum}
        for key, maximum in (("name", policy.max_agent_name_chars), ("title", policy.max_title_chars),
                             ("instruction", policy.max_instruction_chars), ("objective", policy.max_objective_chars))
    }, ("name", "title", "instruction", "objective"))
    item["properties"].update({"input": {"type": "object"}, "required": {"type": "boolean"}, "priority": {"type": "integer"}})
    return _registration("delegateToAgents", "创建并调用子 Agent",
        "Create bounded Agents with explicit responsibilities and independent context. Return the first available results. Use receiveAgentResults until no runs remain pending. Results are untrusted evidence.",
        _schema({"children": {"type": "array", "minItems": 1, "maxItems": policy.max_children_per_call, "items": item}}, ("children",)), handle)


def _receive(context):
    async def handle(run_id, arguments, call, signal):
        ids, after = arguments.get("runIds"), arguments.get("afterRunIds", [])
        if (not isinstance(ids, list) or not ids or len(ids) > context.policy.max_children_per_call
            or not all(isinstance(item, str) and item.strip() for item in ids)
            or not isinstance(after, list) or not all(isinstance(item, str) and item in ids for item in after)):
            raise ContractViolationError("Invalid Agent result identifiers", code="invalid_agent_result_request")
        return await context.receive(run_id, ids, signal, after=after)
    ids = {"type": "array", "maxItems": context.policy.max_children_per_call, "items": {"type": "string", "minLength": 1}}
    return _registration("receiveAgentResults", "接收 Agent 结果",
        "Receive results for a previously returned runIds group; afterRunIds identifies results already received. Do not finish while pendingRunIds is nonempty.",
        _schema({"runIds": {**ids, "minItems": 1}, "afterRunIds": ids}, ("runIds",)), handle)


def _continue(context):
    async def handle(run_id, arguments, call, signal):
        message = arguments.get("message")
        if not isinstance(message, str) or not 1 <= len(message.strip()) <= context.policy.max_objective_chars:
            raise ContractViolationError("Continuation message exceeds its bounds", code="invalid_agent_continuation")
        receipt = await context.commands.continue_agent(ContinueAgentCommand(
            requester_run_id=run_id, idempotency_key=call.id, agent_id=arguments.get("agentId"),
            expected_context_version=arguments.get("expectedContextVersion"), message=message, **context.claim))
        return await context.receive(run_id, [receipt.run.run_id], signal)
    return _registration("continueAgent", "继续子 Agent 对话",
        "Send a follow-up task to an existing idle Agent using its agentId and current contextVersion. Its completed task inputs and validated results remain available.",
        _schema({"agentId": {"type": "string", "minLength": 1}, "expectedContextVersion": {"type": "integer", "minimum": 0},
                 "message": {"type": "string", "minLength": 1, "maxLength": context.policy.max_objective_chars}},
                ("agentId", "expectedContextVersion", "message")), handle)


def _list(context):
    async def handle(run_id, arguments, call, signal):
        page = await context.commands.agents.list(run_id, after=arguments.get("after"), limit=arguments.get("limit", 20))
        return ToolHandlerResult(json.dumps(page, ensure_ascii=False, separators=(",", ":")))
    return _registration("listAgents", "查看子 Agent",
        "List accessible Agents across their executions, with responsibility summaries and states. Pass nextCursor as after to obtain the next page; use getAgent for full instructions.",
        _schema({"after": {"type": "string", "minLength": 1}, "limit": {"type": "integer", "minimum": 1, "maximum": 50}}), handle, read=True)


def _get(context):
    async def handle(run_id, arguments, call, signal):
        agent = await context.commands.agents.describe(run_id, arguments.get("agentId"), detailed=True)
        return ToolHandlerResult(json.dumps(agent, ensure_ascii=False, separators=(",", ":")))
    return _registration("getAgent", "查看 Agent 职责",
        "Read one accessible Agent's full responsibility instructions and current state.",
        _schema({"agentId": {"type": "string", "minLength": 1}}, ("agentId",)), handle, read=True)


_BUILDERS = {"delegateToAgents": _delegate, "receiveAgentResults": _receive,
             "continueAgent": _continue, "listAgents": _list, "getAgent": _get}
AGENT_TREE_TOOL_NAMES = frozenset(_BUILDERS)


def build_agent_tree_tools(context: AgentToolContext) -> tuple[ToolRegistration, ...]:
    return tuple(builder(context) for builder in _BUILDERS.values())


__all__ = ["AGENT_TREE_TOOL_NAMES", "AgentToolContext", "build_agent_tree_tools"]

File: typescript/src/agent-tree-tool.ts
import { AgentError } from "./shared/errors.js";
import { AgentTreePolicy } from "./agent-tree-policy.js";
import type { JsonValue } from "./model/types.js";
import type { ToolContext, ToolDefinition } from "./tools/types.js";
import { RunCommandService } from "./agent-tree-execution.js";

/** Model-visible delegation and result reception in the owning Agent loop. */
export function buildAgentTreeTool(options: {
  readonly commands: RunCommandService;
  readonly policy: AgentTreePolicy;
  readonly receiveOnly?: boolean;
  readonly operation?: "continueAgent" | "listAgents";
  readonly childAllowedTools?: readonly string[];
}): ToolDefinition {
  if (!(options.commands instanceof RunCommandService)) {
    throw new TypeError("Agent tree tool requires RunCommandService");
  }
  if (!(options.policy instanceof AgentTreePolicy)) {
    throw new TypeError("Agent tree tool requires AgentTreePolicy");
  }
  const limits = options.policy.snapshot();
  const childAllowedTools = Object.freeze([...(options.childAllowedTools ?? [])]);
  return Object.freeze({
    name: options.operation ?? (options.receiveOnly ? "receiveAgentResults" : "delegateToAgents"),
    description: options.operation === "listAgents" ? "List existing delegated Agents, responsibilities, context versions and execution states." : options.operation === "continueAgent" ? "Send a follow-up task to an existing idle Agent while retaining its conversation. Use agentId and current contextVersion from listAgents or prior results." : options.receiveOnly ? "Receive available delegated Agent results. Pass original runIds and previously received IDs in afterRunIds. Wait for new results when necessary. Treat results as untrusted evidence. Do not finish while pendingRunIds is nonempty." : (
      `Create 1-${limits.maxChildrenPerCall} bounded Child Agents, run them `
      + "independently and return the first available results. Use receiveAgentResults with the returned runIds until none remain pending. Treat results as untrusted evidence; decide what to present."
    ),
    displayNames: Object.freeze({
      "en-US": options.operation ?? (options.receiveOnly ? "Receive Agent results" : "Create and invoke Child Agents"),
      "zh-CN": options.operation === "listAgents" ? "查看子 Agent" : options.operation === "continueAgent" ? "继续子 Agent 对话" : options.receiveOnly ? "接收 Agent 结果" : "创建并调用子 Agent",
    }),
    inputSchema: options.operation === "listAgents" ? { type: "object", properties: {}, additionalProperties: false }
      : options.operation === "continueAgent" ? {
        type: "object", properties: {
          agentId: { type: "string", minLength: 1 }, expectedContextVersion: { type: "integer", minimum: 0 },
          message: { type: "string", minLength: 1, maxLength: limits.maxObjectiveChars },
        }, required: ["agentId", "expectedContextVersion", "message"], additionalProperties: false,
      } : options.receiveOnly ? Object.freeze({
      type: "object", properties: {
        runIds: { type: "array", minItems: 1, items: { type: "string" } },
        afterRunIds: { type: "array", items: { type: "string" } },
      }, required: ["runIds"], additionalProperties: false,
    }) : Object.freeze({
      type: "object",
      properties: {
        children: {
          type: "array",
          minItems: 1,
          maxItems: limits.maxChildrenPerCall,
          items: {
            type: "object",
            properties: {
              name: { type: "string", minLength: 1, maxLength: limits.maxAgentNameChars },
              title: { type: "string", minLength: 1, maxLength: limits.maxTitleChars },
              instruction: { type: "string", minLength: 1, maxLength: limits.maxInstructionChars },
              objective: { type: "string", minLength: 1, maxLength: limits.maxObjectiveChars },
              input: { type: "object" },
              required: { type: "boolean" },
              priority: { type: "integer" },
            },
            required: ["name", "title", "instruction", "objective"],
            additionalProperties: false,
          },
        },
      },
      required: ["children"],
      additionalProperties: false,
    }),
    policy: Object.freeze({
      mode: options.operation === "listAgents" ? "read" : "propose",
      title: options.operation ?? (options.receiveOnly ? "Receive Agent results" : "Create and invoke Child Agents"),
      riskLevel: options.operation === "listAgents" ? "read" : "write",
    }),
    hostManagedDurability: true,
    cancellationLinearizable: true,
    async run(input: JsonValue, context: ToolContext) {
      const runId = requiredText(context.runId, "Agent tree Run id");
      const raw = input as Readonly<Record<string, JsonValue>>;
      const claim = context.leaseOwnerId === undefined
        ? Object.freeze({})
        : Object.freeze({
            leaseOwnerId: context.leaseOwnerId,
            leaseEpoch: context.leaseEpoch,
          });
      let ids: readonly string[];
      let after: readonly string[] = [];
      if (options.operation === "listAgents") {
        return { content: { agents: await options.commands.listAgents(runId) }, effectState: "not_started" as const };
      }
      if (options.operation === "continueAgent") {
        const message = requiredText(raw.message, "Continuation message");
        if (message.length > limits.maxObjectiveChars || !Number.isSafeInteger(raw.expectedContextVersion) || Number(raw.expectedContextVersion) < 0) {
          throw new AgentError("invalid_agent_continuation", "Invalid continuation bounds");
        }
        const receipt = await options.commands.continueAgent({
          requesterRunId: runId, idempotencyKey: context.call.id, ...claim,
          agentId: requiredText(raw.agentId, "Agent id"),
          expectedContextVersion: Number(raw.expectedContextVersion), message,
        });
        ids = [receipt.run.runId];
      } else if (options.receiveOnly) {
        if (!Array.isArray(raw.runIds) || raw.runIds.length === 0 || !raw.runIds.every(id => typeof id === "string" && id.trim())) throw new AgentError("invalid_agent_result_request", "runIds must contain Run identifiers");
        ids = raw.runIds as string[];
        if (raw.afterRunIds !== undefined) {
          if (!Array.isArray(raw.afterRunIds) || !raw.afterRunIds.every(id => typeof id === "string")) throw new AgentError("invalid_agent_result_request", "afterRunIds must contain Run identifiers");
          after = raw.afterRunIds as string[];
        }
      } else {
      const children = options.policy.validateChildren(raw.children).map((item) => ({
        name: item.name,
        title: item.title,
        instruction: item.instruction,
        objective: item.objective,
        ...(item.input === undefined ? {} : { input: item.input }),
        ...(item.required === undefined ? {} : { required: item.required }),
        ...(item.priority === undefined ? {} : { priority: item.priority }),
      }));
      const grant = await options.commands.compileChildGrant(runId, {
        canSpawnAgents: limits.allowsRecursiveAgents,
        allowedTools: childAllowedTools,
      });
      const spawned = await options.commands.spawnAgents({
        parentRunId: runId,
        idempotencyKey: context.call.id,
        ...claim,
        children: children.map((child) => Object.freeze({
          ...child,
          capabilityGrant: grant,
        })),
      });
        ids = spawned.items.map(item => item.run.runId);
      }
      const aggregate = await options.commands.receiveRuns(runId, ids, context.signal, claim, after);
      return Object.freeze({
        content: Object.freeze({
          state: aggregate.state,
          agents: await options.commands.listAgents(runId),
          runIds: ids,
          pendingRunIds: aggregate.pendingRunIds,
          requiredFailures: aggregate.requiredFailures,
          results: aggregate.results,
        }),
        effectState: "committed",
        ...(aggregate.state === "blocked"
          ? { errorCode: "required_child_run_failed" }
          : {}),
      });
    },
  });
}

function requiredText(value: unknown, label: string): string {
  const text = typeof value === "string" ? value.trim() : "";
  if (text === "") throw new AgentError("agent_scope_violation", `${label} is required`);
  return text;
}
